### ISSUE TITLE
Add support for adding/removing aria and other attributes as validation is added/removed in the code when using the Reactive Forms (previously model-driven forms) approach

### ISSUE THREAD

**Issue by DeborahK** [angular/angular#9121]
- **I'm submitting a ...**
- [ ] bug report
- [x] feature request
- [ ] support request => Please do not submit support request here, see note at the top of this template.

**Current behavior**
When using the reactive forms/model-driven approach, the validation rules for HTML input elements are defined in the component class, not the HTML as attributes. This is great in that the validation can be tailored to different user types or application state. For example, a field is required for normal users but not required for admin users.

However, we lose some built-in browser functionality when these attributes are not defined in the HTML, such as limiting character entry (maxlength) and supporting accessibility (required or aria-required)

**Expected/desired behavior**
For built-in validation types, provide the option to add/remove  attributes automatically. For example, adding the Validator.required to a control in the class should automatically add the required attribute to the DOM.

For custom validation, provide a way for the validation author to specify the attributes to add/remove from the DOM.
- **Please tell us about your environment:**
- **Angular version:** 2.0.0-RC 
- **Browser:** all 
- **Language:** all

**Comment by ShiftySituation** [comment]
Do we really want to rely on browser implementations for validation?

**Comment by DeborahK** [comment]
That is certainly not what I meant. But for screen readers to understand a required field, for example, the required attribute needs to be on the input element.
So leave validation operationally as it is now ... but add the input element attributes for screen readers and such.

**Comment by AlmeroSteyn** [comment]
To summarize what we need for accessibility.

Firstly the standard validator attributes, such as `pattern`, `required`, `minlength` etc. These should be added by the framework to the form controls to support browser implementation, as mentioned above, as well as accessibility. 

Then we need to properly decorate with the `ARIA` states and properties.

`aria-required`: For backwards compatibility this should be added if the field is required, but in HTML5 we can simply decorate with the standard 'required`.

`aria-invalid`: True or false based on the current validity.

`aria-describedby`: This needs to contain, at any moment, the ID or list of IDs of any visible elements that further describe the decorated form control. So this would include descriptor fields as well as any field level validation error messages we display. This needs to be managed dynamically during run time as having any ID's in this list not present on the page, even if it is hidden, creates accessibility violations and invalid HTML.

So, with that said I would think that the first set should be added by the framework but I am wondering if the `ARIA`  tags should not be added by the developer.

The complexity of `aria-describedby` alone, with having to map it to ID's that may even be autogenerated inside components, would make me think that it is not a framework issue.

Also, developers would need to do exactly the same work on *template driven forms*  to make them accessible. 

So I'd think it should be added for both, if it is added at all. But then it opens questions about the other `ARIA` attributes for other functions and we are heading into `ng-aria` waters again.

**Comment by zoechi** [comment]
What about a distinct aria directive (with its own inputs for configuration) that fetches information from the control and updates the attributes?

**Comment by mohammedzamakhan** [comment]
Any updates on this?

**Comment by kyubisation** [comment]
I gave some thought to this issue and came up with the following proposal.
The idea is based around the Reflect Metadata API.
I would appreciate your feedback.

## Proposal

### Add interpretable metadata to the validators

Use Reflect to add metadata to the validators. The metadata would be similar to the error object returned, when the validation fails.
Validators.compose (and also Validators.composeAsync) would aggregate the metadata of the given validators and define the metadata of the merged validator.

e.g.:
```ts
  /**
   * Validator that requires controls to have a non-empty value.
   */
  @AccessibilityMeta({'required': true})
  static required(control: AbstractControl): ValidationErrors|null {
    return isEmptyInputValue(control.value) ? {'required': true} : null;
  }

...

  /**
   * Validator that requires controls to have a value of a maximum length.
   */
  static maxLength(maxLength: number): ValidatorFn {
    const validator = (control: AbstractControl): ValidationErrors | null => {
      ...
    };
    defineAccessibilityMeta(validator, {'maxlength': maxLength});
    return validator;
  }
  
...

  /**
   * Compose multiple validators into a single function that returns the union
   * of the individual error maps.
   */
  static compose(validators: null): null;
  static compose(validators: (ValidatorFn|null|undefined)[]): ValidatorFn|null;
  static compose(validators: (ValidatorFn|null|undefined)[]|null): ValidatorFn|null {
    ...

    const mergedAccessibilityMeta = aggregateAccessibilityMeta(presentValidators);
    const validator = function(control: AbstractControl) {
      return _mergeErrors(_executeValidators(control, presentValidators));
    };
    defineAccessibilityMeta(validator, mergedAccessibilityMeta);
    return validator;
  }
```

### Create accessibility interpreter service

Implement a service, which maps the metadata to aria and validation attributes and also compares them with a whitelist of allowed attributes, based on input type.

e.g.:
{'required': true} -> required, aria-required
{'maxlength': 15} -> maxlength="15"

input[type=text]
{'required': true, 'min': 3} -> required, aria-required // 'min' has no purpose for text inputs

### Create directive for form control accessibility

Implement a new directive to render the appropriate attributes. This directive resolves the FormControl instance and the accessibility interpreter service.

#### a) Define new selector (formControlAccessibilty) in order to allow an opt-in scenario

e.g.:
```ts
@Directive({
  selector: 'input[formControlAccessibilty],textarea[formControlAccessibilty]',
})
export class FormAccessiblityDirective {
  @Input() type: string;
  
  ...
}
```

#### b) Reuse existing form control selectors

This could be added to its own module (e.g. FormsAccessiblityModule), together with the service, in order to make it opt-in.

e.g.:
```ts
@Directive({
  selector: 'input[formControlName],textarea[formControlName],input[formControl],textarea[formControl]',
})
export class FormAccessiblityDirective {
  @Input() type: string;
  
  ...
}
```

**Comment by kyubisation** [comment]
@kara I would like to implement this as a pull request, but I would like some feedback first, in order to figure out if this is an acceptable proposal.

**Comment by Da13Harris** [comment]
Per the summary from @AlmeroSteyn :
> aria-invalid: True or false based on the current validity.

I wonder if this portion can be achieved reasonably quickly, since it seems to be low hanging fruit.

**Comment by kyubisation** [comment]
Probably. A basic implementation would be as follows: https://stackblitz.com/edit/angular-obkdtf
I apologize for this quick and dirty example. I currently lack the time to publish this as part of a package.

**Comment by stefdelec** [comment]
@kyubisation I like it so far. However you may want to add a check if a formControl exist or not.
I can create a repo on git from your work, if you allow me to contribute (and copy--past !).
In my current project we need to be WCAG AA compliant.

**Comment by kyubisation** [comment]
Feel free to copy and modify this code. It was merely intended as an example.

**Comment by twerske** [comment]
Thanks for submitting the proposal! 
We've discussed a11y improvements for the Forms package logic and the conclusion was that adding `aria` attributes as a part of the Forms validation would be a breaking change, so it'd not be easy to land it. Angular Forms logic is used in many apps and lib where developers might already have `aria` attributes setup in templates, so Forms logic might conflict (override) some of the values. We'll be investigating other ways to improve a11y (including a possibility of integrating some a11y lint tools) and share more info once we have it. For now I'm going to close this ticket as no action is required from the Forms package perspective.
